### ISSUE TITLE
Can't create UMD build which can be required by Node

### ISSUE THREAD

**Issue by donaldpipowitch** [webpack/webpack#6522]
**Do you want to request a *feature* or report a *bug*?**

*bug*

**What is the current behavior?**

If you require a module in the target output UMD you get this error:

```
ReferenceError: window is not defined
    at Object.<anonymous> (/mnt/c/Users/foo/webpack-umd-node/dist/main.js:1:272)
```

**If the current behavior is a bug, please provide the steps to reproduce.**

* https://github.com/donaldpipowitch/webpack-umd-node
* `$ git clone git@github.com:donaldpipowitch/webpack-umd-node.git && cd webpack-umd-node && yarn && yarn build && yarn test`

**What is the expected behavior?**

You shouldn't see a `ReferenceError`.

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

Node v8.9.4, webpack v4.0.0-beta.2

**Comment by sokra** [comment]
At which line does the error happen?

**Comment by sokra** [comment]
hmm... I see.

Try `target: "node"` or `output.globalObject: "xxx"`

**Comment by donaldpipowitch** [comment]
> output.globalObject: "xxx"

Is this a new setting? Haven't seen it so far. What should the value be in the case of a universal module?

**Comment by donaldpipowitch** [comment]
With `target: 'node'` it throws `Uncaught ReferenceError: global is not defined` if the UMD build is loaded in the browser 😄

**Comment by sokra** [comment]
Maybe that's the better solution: #6525

**Comment by emilio-martinez** [comment]
Just ran into this issue while trying to upgrade to Webpack 4.x—I have a lib that is for any env. Looking at my previous releases, `window` used to be in fact `this`.

@sokra when I was looking at what the defaults are, I noticed that there's a `default` for `self`; however, because of the config validation on `target`, you can't just place anything. Was your intent to make `self` the "universal" option?

https://github.com/webpack/webpack/blob/213226ece261a0e89edb9de15fbc6775418f6041/lib/WebpackOptionsDefaulter.js#L114-L129

Just wondering. In the meantime, `globalObject: 'this'` seems to work.

**Comment by emilio-martinez** [comment]
It looks like `self` is `Window` in browserland but is `undefined` in `node` anyway. Not sure if `this` is the global for workers.

**Comment by sokra** [comment]
We had to switch to `this` to something else, because `this` is `undefined` with `<script type="module">`

**Comment by bhovhannes** [comment]
Setting `globalObject: 'typeof self !== \'undefined\' ? self : this'` produces exactly the same UMD header as in WebPack 3

**Comment by ScottRudiger** [comment]
> Setting `globalObject: 'typeof self !== \'undefined\' ? self : this'` produces exactly the same UMD header as in WebPack 3

Thanks, at first I thought this in combination with setting `libraryExport: 'default'` allowed me to require my modules as expected with CommonJS.

However, I see quite strange, seemingly random behavior with the following actions:

1) run `webpack --mode development` 
    - builds appropriately; bundle.js shows:
```js
...
})(typeof self !== 'undefined' ? self : this, function() {
...
```
2) save a file without any changes being made:
    - bundle.js reverts to the following without even running `webpack --mode development` again:
```js
...
})(window, function() {
...
```
3) run `webpack --mode development` again
    - builds appropriately
4) make a change and save file without building again
    - reverts to `window` again

Similarly random behavior can be seen by running `webpack --mode development --watch`, then saving a file with or without making any changes. One of the following 2 results occur:
1) builds appropriately
or
2) reverts to `window`

The strange thing is that when saving a file without any changes being made, bundle.js is not even rebuilt; but still, the line `})(typeof self !== 'undefined' ? self : this, function() {` randomly changes back and forth to `})(window, function() {`.

One thing I tried was deleting `node_modules/.cache/babel-loader` (as I was setting cacheDirectory to true in development mode). But turning that setting off and deleting `.cache/` entirely made no change to the above behavior.

Should I create a separate issue for this, or does it seem related? Let me know if more environment details are needed.

**Comment by aldarund** [comment]
// Rant mode on
That hilarious, as i see it was in 4 beta already. Now its 4.8.3, and its still there without a proper fix. Spent a day for this. Behaviour so inconsistent that i wanted to break something xD So many combinations and only one really worked for me.
// Rant mode off
Proposed solution  from above ( globalObject: 'typeof self !== \'undefined\' ? self : this' ) simply dont work for me in browser. 
my lib
```
const LibraryModule = {
  x: 'y'
}
export default LibraryModule;
```
my import logging code
```
import * as x from "mylib"
console.log('import', x)
```
On server it as expected produce
``` import { default: { x: 'y' } } ```
In browser it produce
``` import {__esModule: true} ```
So empty module in browser for some reason.
If i set
``` globalObject: 'this',```
It almost worked. On server it was fine like in previous case.
But in browser it was
```
import {mylib: {default: { x: 'y' } }, __esModule: true}
```
Basically it was adding for some reason mylib in browser. But not in node. Really consistent....
And only removing from config
```  library: "mylib", ```
made it behave same for me in browser and node.

TLDR
ONLY
globalObject: 'this' AND removing option from config library: "mylib" made it work same in browser and node for me. 
And im not really sure yet if it will work properly or no. It just feels so hacky and inconsistent that it might break anytime.

**Comment by aldarund** [comment]
Ok, even last solution not fully worked. Well, it seems to be working but im still getting error messages in both browser and node console that
``export 'default' (imported as 'mylib') was not found in 'mylib'``
It happens if i import normally 
```import  mylib from "mylib"```
if i import it like in prev message error doesnt happen
```import * as mylib from "mylib"```
And it happens after it print out module content that is a proper one in browser, and before proper content in node. And later it seems to work and accessible despite this errors...
This is killing me..

**Comment by ScottRudiger** [comment]
I'm not sure Webpack actually provides a method of outputting ES modules for the browser. I've probably spent a cumulative day searching for how it could be done. The conclusion I came to is it's not possible at this time. You have at least a couple alternatives: 1) output modules with babel 2) use rollup.

Someone _**please**_ correct me if I'm wrong.

Since my lib is already set up with Webpack (minus ES modules) I'll probably generate them with babel for the time being, and (maybe someday) refactor once Webpack supports this or do a big refactor to rollup.

**Comment by nebrelbug** [comment]
So... any updates? I need my library to work with CommonJS. Right now, with `libraryTarget: 'umd'`, it works in the browser but not with CommonJS.

**Comment by yairEO** [comment]
Why did anyone wrote:

    })(window, function() {

when it should be:

   })(this, function() {

It's pretty obvious no?

**Comment by nebrelbug** [comment]
@zacharyfmarion @yairEO I found a solution; add the following line to your webpack.config.js.
```globalObject: "typeof self !== 'undefined' ? self : this"```.
Kind of frustrating that webpack hasn't fixed the bug yet though.

**Comment by tannermeans** [comment]
To add to this, we ran into an issue combining optimization.splitChunks with the targetObject.
To get around this, we added parens around the suggestion from @bhovhannes (https://github.com/webpack/webpack/issues/6522#issuecomment-371120689):

```
  output: {
    ... configuration for multiple entry points
    libraryTarget: 'umd',
    globalObject: "(typeof window !== 'undefined' ? window : this)",
  },
  optimizations: {
    splitChunks: {
      cacheGroups: {
        base: {
          name: 'common-chunk',
          chunks: 'initial',
          minChunks: 2,
        },
      },
    },
  }
````
Our situation might be unique, but we are creating a shared library where we want to ship N components, along with a common chunk to provide at build-time with shared code between the components.

e.g. 
```
import '@our_ns/our-lib/common-chunk`
import '@our_ns/our-lib/Component1`
import '@our_ns/our-lib/Component2`
```

_not_ having parens around `(typeof window !== 'undefined' ? window : this)` resulted in the following code in the commons chunk:

```
typeof self !== 'undefined' ? window : this["webpackJsonp"] = typeof self !== 'undefined' ? window : this["webpackJsonp"] || []).push([["common-chunk"]
```
Clearly in the browser that first conditional will be false `typeof self === 'object'` (self === window), causing the rest of the ternary op to be ignored and `window.push` to be run.

Adding parens around the globalObject outputs this code:
```
((typeof window !== 'undefined' ? window : this)["webpackJsonp"] = (typeof window !== 'undefined' ? window : this)["webpackJsonp"] || []).push([["common-chunk"],{
```

It's a hack that really depends heavily on the [internals of webpack](https://github.com/webpack/webpack/blob/d5983ed62b1bcf0b5b3e8634aa243b055357b83a/lib/web/JsonpMainTemplatePlugin.js#L466), but I don't see another way around it until this is solved.

Sorry if this doesn't work for you, this is DEFINTELY a "works on my machine" kinda solution

**Comment by geronimus** [comment]
Since the issue has been open for nearly a year, it does not appear a fix will be forthcoming. In that case, perhaps the solution is to add the work-arounds mentioned here to the documentation pages:
- [Output](https://webpack.js.org/configuration/output/#output-librarytarget)
- [Authoring Libraries](https://webpack.js.org/guides/author-libraries)

**Comment by aldegoeij** [comment]
referencing tickets #3486, #3506, #3615, #3015, #876, #3477 and webpack/webpack#6522

**Comment by vbabak** [comment]
According to the [docs](https://webpack.js.org/configuration/output/#outputglobalobject) include `output.globalObject` into **webpack.config.js**:

```
module.exports = {
  output: {
    libraryTarget: 'umd',
    globalObject: 'this'
  }
}
```
_To make UMD build available on both browsers and Node.js, set output.globalObject option to 'this'._

**Comment by dwiyatci** [comment]
> According to the [docs](https://webpack.js.org/configuration/output/#outputglobalobject) include `output.globalObject` into **webpack.config.js**:
> 
> ```
> module.exports = {
>   output: {
>     libraryTarget: 'umd',
>     globalObject: 'this'
>   }
> }
> ```
> 
> _To make UMD build available on both browsers and Node.js, set output.globalObject option to 'this'._

@vbabak This is not working anymore somehow once I add:
```javascript
optimization: { 
  splitChunks: { chunks: 'all' } 
}
``` 
to my webpack config 🙁

**Comment by vbabak** [comment]
> > According to the [docs](https://webpack.js.org/configuration/output/#outputglobalobject) include `output.globalObject` into **webpack.config.js**:
> > ```
> > module.exports = {
> >   output: {
> >     libraryTarget: 'umd',
> >     globalObject: 'this'
> >   }
> > }
> > ```
> > 
> > 
> > _To make UMD build available on both browsers and Node.js, set output.globalObject option to 'this'._
> 
> @vbabak This is not working anymore somehow once I add:
> 
> ```js
> optimization: { 
>   runtimeChunk: true 
> }
> ```
> 
> to my webpack config 🙁

@dwiyatci I'm not a maintainer. I had shared what helped me at that moment.

**Comment by dwiyatci** [comment]
@vbabak I knew. Was just gonna leave it for future ref. 🙂

**Comment by aendrew** [comment]
That bit in the documentation is super key but really hidden. Can perhaps this issue be changed to a documentation issue about needing a warning for NodeJS targets if UMD is used without setting globalObject to `this`?

**Comment by alexander-akait** [comment]
@aendrew We are based on browserslist by default, if you support `globalThis` by versions, we will use `globalThis` and it will work everywhere

**Comment by mihanizm56** [comment]
any updates?
migrate from webpack 4 to 5 ejected CRA app - and if do "globalObject: 'this'" then my build is broken
